### DISCUSSION TITLE
Улучшение onlyFields у FormElements::class

### DISCUSSION THREAD

**Discussion by MadZee-M** [moonshine-software/moonshine#1127] [Ideas]
Доброго времени суток!

Предлагаю улучшить метод onlyFields у класса FormElements.
На данный момент, если добавить поле StackFields, то onlyFields не получает все дочерние поля у StackFields.

Предлагаю добавить новый аргумент "**flatten**"

1. `public function onlyFields(bool $withWrappers = false, $flatten = false): Fields {}`
2. Изменить метод
 ```
public function onlyFields(bool $withWrappers = false, $flatten = false): Fields
    {
       ...
        $flatten
            ? $this->extractFlattenFields($this->toArray(), $data)
            : $this->extractFields($this->toArray(), $data);
       ...
    }
```
3. Добавить новый метод "**extractFlattenFields**"

```
    protected function extractFlattenFields($elements, array &$data): void
    {
        foreach ($elements as $element) {
            if ($element instanceof Field) {
                $data[] = $element;
            }
            
            if ($element instanceof HasFields) {
                $this->extractFields($element->getFields(), $data);
            }
        }
    }
```

Таким образом, в случае, если требуется получить все поля, то не возникнет проблем.

**Comment by lee-to** [comment]
onlyFields исключает все что не поля и если там StackFields то у StackFields ->getFields будет получать дочернии, flatten нам не нужен сейчас в коробке поэтому и добавлять его как мертвый код мы не будем, но вы можете сделать свой декоратор или через macroable самостоятельно
